### ISSUE TITLE
Changing validation error messages at runtime

### ISSUE THREAD

**Issue by makepanic** [DockYard/ember-changeset-validations#94]
Hi, great addon.

I'm wondering, if there is a recommended way of changing validation messages at runtime?

I know it's possible to change it initially by defining a module that matches [`/validations\/messages$/gi`](https://github.com/DockYard/ember-changeset-validations/blob/master/addon/utils/get-messages.js#L7) but it's only loaded once. 

```js
define(['foo/validations/messages'], () => {
  return {default: {
     ...messages
  }};
})
```

If a user changes its locale, it's impossible to update validation messages from what i can tell.
Are there any known workarounds for supporting this kind of use-case?

**Comment by martndemus** [comment]
@makepanic Are you using ember-i18n or equivalent?

**Comment by makepanic** [comment]
Yes i'm using `ember-intl`.

A configurable service sounds like a great idea.

**Comment by poteto** [comment]
I'm not a fan of the service idea in the addon itself, it would require a lot of rework since most of the addon are vanilla JS modules with no knowledge of Ember. Instead, you can use computed properties on an `Ember.Object`:

```js
const Messages = Ember.Object.extend({
  locale: Ember.inject.service(),

  inclusion: computed('locale', function() {
    // ...
  })
});

export default Messages.create();
```

**Comment by martndemus** [comment]
That would work.

**Comment by makepanic** [comment]
Thanks, 
Using CPs inside the messages object works without problems.

Another question: 
Would it make sense to allow users to pass their own `buildMessage` function to move string template interpolation away from the `validation-errors` module to a pluggable part of the app (primarily i18n related libs)?
That way one isn't coupled to the string format ember-changeset-validations is using.

**Comment by jmacqueen** [comment]
@makepanic How did you get this to work? I can't seem to sort out the manual injection of the (I'm using `ember-i18n`) service using container. Do you have a link to your implementation?

**Comment by jmacqueen** [comment]
Here's what I ended up doing, in case anyone else ends up here looking for help: https://gist.github.com/jmacqueen/6d31aaa5ba5f5841d123d0406f1f86c5

**Comment by makepanic** [comment]
@jmacqueen i went with passing a `buildMessage` option to the `lookupValidator` method (see https://github.com/makepanic/ember-changeset-validations/commit/8cca573dfaebf71c0dec1368d66bdf7cf683eb18).
Message building happens by calling the provided `buildMessage` function which in turn calls ember-intl to translate the given incoming validation message key.

**Comment by celicoo** [comment]
Hei @jmacqueen, i'm seeing this error in my console:

```javascript
Uncaught TypeError: _siteUtilsApplication.default.instance.lookup is not a function
```

Any ideias why?

Thank you! <3

**Comment by karellm** [comment]
I was wondering if this addon could check whether `message.js` export a POJO or an `Ember.Object`. If it is an Ember object, it could take care of the creation and [owner injection](https://offirgolan.github.io/ember-cp-validations/docs/modules/Basic.html#a-note-on-testing--object-containers):

```js
import Ember from 'ember';
const { computed } = Ember;

export default Ember.Object.extend({
  i18n: Ember.inject.service(),

  i18nName: computed('i18n.locale', function () {
    return this.t('validations.i18nName');
  })
});
```

And the addons would be responsible to:

```js
import Messages from 'messages.js';

Messages.create(
  Ember.getOwner(this).ownerInjection()
);
```

Does it sound like it would work? And is it something you'd be willing to implement? cc @poteto @martndemus

**Comment by poteto** [comment]
@karellm I would be open to this. Do you have time to open a PR?

**Comment by karellm** [comment]
@poteto I tried but I'm not very familiar with Ember internals and add-ons development. `this` is actually `undefined` inside the `utils`. It looks like it should be passed by `buildMessage` which makes the API a little weird.

Also I'm not sure how to go about testing this since the resolver looks for `validations/messages` so I can't have 2 different files (one POJO, one Ember.Object).

I'm happy to do it but I would really appreciate some pointers. Thanks!

**Comment by CmdrChaos** [comment]
Any progress on this?  I find myself stuck trying to get an Ember.Object to be recognized form messages.js.  As long as it's a simple POJO - no issues, but when wrapped in an Ember.Object, the add-on reverts to its defaults rather than using the provided object from messages.js.

**Comment by vladaspasic** [comment]
Hi all,

I have just forked this repo in order to implement this feature. @jmacqueen gave a good example on how to connect the Ember IoC with the messages.

Following his advice, I created an instance initializer that would pass the Application instance to the `get-messages.js` file in order to lookup custom messages via Ember container. Using the container there is no need to iterate through all defined modules and addons in order to find the custom `validations/messages.js` file. You can check the fork [here](https://github.com/vladaspasic/ember-changeset-validations)

To use this feature, you would simply need to create your own `instance-initializer` in your application and define it like so:

```
export {default} from 'ember-changeset-validations/instance-initializer/ember-changeset-validations';
```

Then in your `validations/messages.js` file, you can define a property or implement the `messageForKey` function to load a message from either `i18n`, `intl ` or any Ember Service:

```
import Ember from 'ember';
const { get, computed } = Ember;

export default Ember.Object.extend({
  i18n: Ember.inject.service(),

  i18nName: computed('i18n.locale', function () {
    return this.t('validations.i18nName');
  }),

  messageForKey(key) {
    return this.get('i18n').t(`validations.${key}`, {
      'default': get(this, 'defaults.${key}')
    });
  }
});
```
If you are ok with this approach, I would gladly make a PR to incorporate these changes. 


While working on this, I realised that that a simple API could be exposed in order to load custom messages and field descriptions. 
From the current issues it seems that there is need for something similar. It would be great if we could make this library be even more flexible with validation messages.

My suggestions would be to expose `messageForKey` or `descriptionForKey` functions. Message and description lookup would look like this:

```
Ember.Mixin.create({
  // default messages
  defaults: {},

  getDescriptionFor(key = '') {
    let value;

    // Check if there is a custom lookup function defined
    if (canInvoke(this, 'descriptionForKey')) {
      value = this.descriptionForKey(key);
    }
    // Generate default description
    if (isNone(value)) {
      value = capitalize(dasherize(key).split(/[_-]/g).join(' '));
    }
    return value;
  },

  // use the unknown property hook in Ember.Object in order to find
  // messages which were not defined directly on the object
  unknownProperty(key = '') {
    let value;

    // Check if there is a custom lookup function defined
    if (canInvoke(this, 'messageForKey')) {
      value = this.messageForKey(key);
    }
    // Play nice with ObjectProxy instances
    if (isNone(value)) {
      value = this._super(key);
    }
    // Load default message
    if (isNone(value)) {
      value = get(this, `defaults.${key}`);
    }

    return value;
  }
})
```

As this change would be a rather big one, I would be good to agree on this beforehand :)

**Comment by ktrhn** [comment]
Does the solution https://github.com/DockYard/ember-changeset-validations/issues/94#issuecomment-244089956 still work? It seems computed properties aren't merged by `Ember.merge` in https://github.com/DockYard/ember-changeset-validations/blob/c638e81a537c5e9fb427227cc0e35852d01a6919/addon/utils/with-defaults.js#L13-L15

**Comment by hornetnz** [comment]
We implemented the changes @jmacqueen did, and the breakpoint on the initialize method in instance-initializers/application.js is hitting. However nothing is hitting for our validations/messages.js file. Is there something that needs to be added to our validators file to get our translated messages file loaded?

A sample of our validator file...
`
import {
    validateLength,
    validateFormat
} from 'ember-changeset-validations/validators';
import { validatePhone, validateExtension } from 'ember-engine-contacts/validators/phone';
import { regularExpressions } from 'ember-validators/format';
import validateTwitter from 'ember-engine-contacts/validators/twitter';

export default {
    workEmail: [
        validateLength({ max: 254, allowNone: true, allowBlank: true }),
        validateFormat({ type: 'email', allowNone: true, allowBlank: true, regex: regularExpressions.email })
    ]
};
`

I can add 'message' to the object parameters in the validateFormat method, which works...but not sure how to get our overridden messages file to work.
Any help would be greatly appreciated!

**Comment by nandorstanko** [comment]
@martndemus @poteto Could you suggest some solution for this issue, guys? I'm using ember-i18n service for my translations, and I've tried the option https://github.com/poteto/ember-changeset-validations/issues/94#issuecomment-244089956 , but this doesn't work anymore. Any other elegant suggestion? Thanks for your great work btw!

**Comment by 0xadada** [comment]
@makepanic if i wanted to use the approach you discussed in the [first post](https://github.com/poteto/ember-changeset-validations/issues/94#issue-174474839), how would I do that, where would i place the file so it hooks and is loaded properly?

**Comment by 0xadada** [comment]
@makepanic @martndemus @karellm @nandorstanko I've released an Ember Addon to apply  `ember-i18n` translations to these messages in a beta version here https://github.com/mirai-audio/ember-i18n-changeset-validations. Could use a few hands to try it before I release a v1.0.0

**Comment by nandorstanko** [comment]
@0xadada Correct me if I'm wrong, but your solution doesn't support changing locale without reloading the app. That's a huge feature, that is a **MUST HAVE** for me... That is why I use ember-i18n addon in the first place. I've switched to ember-changeset-cp-validations+ember-i18n-cp-validations combo, and it's working for me. Probably slower this way, but not an issue so far.

**Comment by bgentry** [comment]
Yesterday, I finally ran into the need to use a service in my validations. In my particular case, I needed to fetch an array of strings from my API to validate inclusion against. After reading this thread, I almost tried to make this work in this addon.

However, I realized that I would essentially be rebuilding a bunch of the stuff in [ember-changeset-cp-validations](https://github.com/offirgolan/ember-changeset-cp-validations). I decided to just start using that addon instead. I think that model is a lot more robust and fits better with any scenario where you need to access dynamic data provided by an Ember service. It was a pretty simple transition and I will probably move my other validations over if the need arises. Definitely recommended!

**Comment by onceatime** [comment]
what you can do is to register a seperate lookup util.

app/utils/lookup-validator.js
```
import originalLookupValidator from 'ember-changeset-validations';
import { merge } from '@ember/polyfills';

export default function lookupValidator(validationMap = {}, options = {}) {
  return ({ key, newValue, oldValue, changes, content }) => {
    const validationResult = originalLookupValidator(validationMap)({ key, newValue, oldValue, changes, content });

    if (!options.intl && !options.intl.t) {
      return validationResult;
    }

    return validationResult.map ? validationResult.map(({ message, attrs }) => {
      return options.intl.t(message, merge(attrs, { key: options.intl.t(`validations.changeset.keys.${key}`) }));
    }) : validationResult;
  }
}
```

app/validations/messages.js
```
export default {
  inclusion: 'validations.changeset.messages.inclusion',
  exclusion: 'validations.changeset.messages.exclusion',
  invalid: 'validations.changeset.messages.invalid',
  confirmation: "validations.changeset.messages.confirmation",
  accepted: 'validations.changeset.messages.accepted',
  empty: "validations.changeset.messages.empty",
  blank: 'validations.changeset.messages.blank',
  present: "validations.changeset.messages.present",
  collection: 'validations.changeset.messages.collection',
  singular: "validations.changeset.messages.singular",
  tooLong: 'validations.changeset.messages.tooLong',
  tooShort: 'validations.changeset.messages.tooShort',
  between: 'validations.changeset.messages.between',
  before: 'validations.changeset.messages.before',
  onOrBefore: 'validations.changeset.messages.onOrBefore',
  after: 'validations.changeset.messages.after',
  onOrAfter: 'validations.changeset.messages.onOrAfter',
  wrongDateFormat: 'validations.changeset.messages.wrongDateFormat',
  wrongLength: 'validations.changeset.messages.wrongLength',
  notANumber: 'validations.changeset.messages.notANumber',
  notAnInteger: 'validations.changeset.messages.notAnInteger',
  greaterThan: 'validations.changeset.messages.greaterThan',
  greaterThanOrEqualTo: 'validations.changeset.messages.greaterThanOrEqualTo',
  equalTo: 'validations.changeset.messages.equalTo',
  lessThan: 'validations.changeset.messages.lessThan',
  lessThanOrEqualTo: 'validations.changeset.messages.lessThanOrEqualTo',
  otherThan: 'validations.changeset.messages.otherThan',
  odd: 'validations.changeset.messages.odd',
  even: 'validations.changeset.messages.even',
  positive: 'validations.changeset.messages.positive',
  multipleOf: 'validations.changeset.messages.multipleOf',
  date: 'validations.changeset.messages.date',
  email: 'validations.changeset.messages.email',
  phone: 'validations.changeset.messages.phone',
  url: 'validations.changeset.messages.url',

  formatMessage: (message, attrs) => {
    return { message, attrs };
  }
}
```

/translations/de-de.yml
```
##############################
# VALIDATIONS
##############################
validations:
  changeset:
    messages:
      inclusion: '{key} is not included in the list'
      exclusion: '{key} is reserved'
      invalid: '{key} is invalid'
      confirmation: "{key} doesn't match {on}"
      accepted: '{key} must be accepted'
      empty: "{key} can't be empty"
      blank: '{key} must be blank'
      present: "{key} can't be blank"
      collection: '{key} must be a collection'
      singular: "{key} can't be a collection"
      tooLong: '{key} is too long (maximum is {max} characters)'
      tooShort: '{key} is too kurz (minimum is {min} characters)'
      between: '{key} must be between {min} and {max} characters'
      before: '{key} must be before {before}'
      onOrBefore: '{key} must be on or before {onOrBefore}'
      after: '{key} must be after {after}'
      onOrAfter: '{key} must be on or after {onOrAfter}'
      wrongDateFormat: '{key} must be in the format of {format}'
      wrongLength: '{key} is the wrong length (should be {is} characters)'
      notANumber: '{key} must be a number'
      notAnInteger: '{key} must be an integer'
      greaterThan: '{key} must be greater than {gt}'
      greaterThanOrEqualTo: '{key} must be greater than or equal to {gte}'
      equalTo: '{key} must be equal to {is}'
      lessThan: '{key} must be less than {lt}'
      lessThanOrEqualTo: '{key} must be less than or equal to {lte}'
      otherThan: '{key} must be other than {value}'
      odd: '{key} must be odd'
      even: '{key} must be even'
      positive: '{key} must be positive'
      multipleOf: '{key} must be a multiple of {multipleOf}'
      date: '{key} must be a valid date'
      email: '{key} ist keine valide email'
      phone: '{key} must be a valid phone number'
      url: '{key} must be a valid url'
    keys:
      email: E-Mail
      password: Passwort
```

and finally use it....
```
...
import lookupValidator from '../../../utils/lookup-validator';
...
return new Changeset({}, lookupValidator(UserValidations, {intl: this.intl}), UserValidations);
```

**Comment by xcambar** [comment]
Hi,
I've come up with a PR that allows to postpone i18n/intl work outside of ember-changeset-validations by making it return (optionally) an object instead of a String. It works wonders for us: https://github.com/poteto/ember-changeset-validations/pull/185

**Comment by scottkidder** [comment]
I can confirm that @xcambar's solution is very flexible and works great for my use case.

**Comment by xcambar** [comment]
FYI, #185 has been merged.

See [the PR](#185) and [this setion of the docs](https://github.com/poteto/ember-changeset-validations/#raw-error-output) for info.

Should this issue be closed now?

**Comment by makepanic** [comment]
Looks good. Can we maybe add an example on how to use this with intl/i18n?

**Comment by xcambar** [comment]
I'm using it with something of the form (wrapped in a helper):

```
{{t (concat "path.to.errors." error.type) errors.context}}
```

**Comment by scottkidder** [comment]
I have kind of tackled this and our migration from `ember-i18n` to `ember-intl` at the same time. The one gotcha I ran into was that the `{{t }}` helper from `ember-i18n` allowed passing a hash of i18n paramaters, where the one from `ember-intl` requires passing each as it's own parameter, which isn't ideal when we don't know the parameters upfront.

**Comment by gauravjain028** [comment]
@xcambar passing `errors.context` with name attribute is not working on my end ? Can you help with that ?

**Comment by xcambar** [comment]
I can try :)
In order to ease investigation, can you share a [twiddle](https://ember-twiddle.com/) or a repo that reproduces the behaviour, please?

**Comment by gauravjain028** [comment]
@xcambar I have setup it on https://github.com/gauravjain028/ember-intl-changeset-error-handling
Template File : https://github.com/gauravjain028/ember-intl-changeset-error-handling/blob/master/app/templates/application.hbs
When in load the app, it give me js error
`The intl string context variable 'description' was not provided to the string '{description} can't be blank'`

**Comment by k-dauda** [comment]
I was having issues with the ember octane but I was able to get a good work around with custom validators.

```
import Component from '@glimmer/component';
import Changeset from 'ember-changeset';
import lookupValidator from 'ember-changeset-validations';
import  { inject as service } from '@ember/service';

export default class MyComponent extends Component {
  @service intl;

  validator({ errorKey }) {
    let errorMsg = this.intl.t(`${errorKey}.error.msg`);

    return function(_, value) {
      let isInValid = ....; // custom validation logic

      if (isInValid) {
        return errorMsg;
      }

      return true;
    }
  }

  validators = {
    name: this.validator({  errorKey: 'name.field' }).
    email: this.validator({ errorKey: 'email.field' })
  };

  changeset = new Changeset(
    { 
      name: this.name,
      email: this.email
    },
    lookupValidator(this.validators),
    this.validators
  );
}
```
